crates/engine/tree/src/tree/mod.rs

```diff
@@ -896,6 +896,26 @@ where
             .remove_persisted_blocks(self.persistence_state.last_persisted_block_number);
     }
 
+    /// Return sealed block from database or in-memory state by hash.
+    fn sealed_header_by_hash(&self, hash: B256) -> ProviderResult<Option<SealedHeader>> {
+        // check database first
+        let block_num = self.provider.block_number(hash)?;
+        if let Some(block_num) = block_num {
+            self.provider.sealed_header(block_num)
+        } else {
+            // Note: it's fine to return the unsealed block because the caller already has
+            // the hash
+            let block = self
+                .state
+                .tree_state
+                .block_by_hash(hash)
+                // TODO: clone for compatibility. should we return an Arc here?
+                .map(|block| block.as_ref().clone().header);
+
+            Ok(block)
+        }
```

Comment:
should we swap these?

Reply:
yeah, actually, checking memory first would be better